flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunction.java

```diff
@@ -0,0 +1,425 @@
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.state.CheckpointListener;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A sink function that collects query results and sends them back to the client.
+ *
+ * <p>This sink works by limiting the number of results buffered in it (can be configured) so
+ * that when the buffer is full, it back-pressures the job until the client consumes some results.
+ *
+ * <p>NOTE: When using this sink, make sure that its parallelism is 1, and make sure that it is used
+ * in a {@link StreamTask}.
+ *
+ * <h2>Communication Protocol Explanation</h2>
+ *
+ * <p>We maintain the following variables in this communication protocol
+ * <ol>
+ *     <li><strong>version</strong>: This variable will be set to a random value when the sink opens.
+ *         Client discovers that the sink has restarted if this variable is different.</li>
+ *     <li><strong>offset</strong>: This indicates that client has successfully received the results
+ *         before this offset. Sink can safely throw these results away.</li>
+ *     <li><strong>lastCheckpointedOffset</strong>:
+ *         This is the value of <code>offset</code> when the checkpoint happens. This value will be
+ *         restored from the checkpoint and set back to <code>offset</code> when the sink restarts.</li>
+ * </ol>
+ *
+ * <p>Client will put <code>version</code> and <code>offset</code> into the request, indicating that
+ * it thinks what the current version is and it has received this much results.
+ *
+ * <p>Sink will check the validity of the request. If <code>version</code> mismatches or <code>offset</code>
+ * is smaller than expected, sink will send back the current <code>version</code> and
+ * <code>lastCheckpointedOffset</code> with an empty result list to indicate an invalid request.
```

Comment:
If client sends a valid offset, but there is currently no new data, we will also send back an empty list. So empty result list not always indicating an invalid request.